reasoner/Reasoner.java

```diff
@@ -63,15 +64,17 @@ public Reasoner(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicMa
     }
 
     public ResourceIterator<ConceptMap> execute(Disjunction disjunction, boolean isParallel) {
-        if (!isParallel) return iterate(disjunction.conjunctions()).flatMap(this::iterator);
-        else return buffer(disjunction.conjunctions().stream()
+        Set<Conjunction> conjunctions = disjunction.conjunctions().stream()
```

Comment:
Here we move type resolving calculation at top level so that it's not nested in the `ResourceIterator::flatMap`. This is because `flatMap` is lazy, and we need to make sure when type resolving happens some one at top level will catch its exception and properly close the transaction. Therefore it cannot be nested in `flatMap`.

Reply:
Moved type resolving to here so that we can deduplicate code for both sync and async case.